### PR TITLE
ARROW-5461: [Java] Add micro-benchmarks for Float8Vector and allocators

### PR DESCRIPTION
**Pull request by liyafan82**
For the past days, we have been involved in some performance related issues. In this process, we have created some performance benchmarks, to help us verify performance results.

Now we want to add such micro-benchmarks to the code base, in the hope that they will be helpful for making performance-related decisions and avoid performance degradation.

### CODE REVIEW

**Comment by emkornfield** [comment]
@liyafan82 thank you for doing this.  Is there a reason you chose to create a new module instead of adding the benchmarks to the existing sub-projects (memory and vector)?

**Comment by liyafan82** [comment]
> @liyafan82 thank you for doing this. Is there a reason you chose to create a new module instead of adding the benchmarks to the existing sub-projects (memory and vector)?

Good question. 

To get reliable and reproducible performance results, JMH is the preferred benchmark framework (as recommended by Jacques Nadeau). And creating a standalone project is the recommended way of doing this. Please see https://openjdk.java.net/projects/code-tools/jmh/

image

File: java/performance/src/main/java/org/apache/arrow/performance/benchmark/micro/Float8Benchmarks.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+   * Tear down benchmarks.
+   */
+  @TearDown
+  public void tearDown() {
+    vector.close();
+    allocator.close();
+  }
+
+  /**
+   * Test reading/writing on {@link Float8Vector}.
```
Comment (emkornfield) [inline java/performance/src/main/java/org/apache/arrow/performance/benchmark/micro/Float8Benchmarks.java]: can you give pointers here to the memory bounds check and null check?

**Review by emkornfield — CHANGES_REQUESTED** [review]
@liyafan82 can the benchmarks be placed under test instead of main?

Also would it make sense to put them in the same package as the classes they are testing, I think that would make them more discoverable?

**Comment by liyafan82** [comment]
> @liyafan82 can the benchmarks be placed under test instead of main?
> 
> Also would it make sense to put them in the same package as the classes they are testing, I think that would make them more discoverable?

I have moved them to the same packages as their related classes. Thanks for the good suggestion.

In addition, I have also placed them as test cases in the test directory. However, I am not sure if this is exactly what you want, since benchmark evaluating can be time consuming. Do you want these benchmarks to be moved to the test directory, and still evaluated by the main method?

File: java/performance/src/main/java/org/apache/arrow/performance/benchmark/micro/Float8Benchmarks.java (hunk 1 above)
Comment (liyafan82) [reply java/performance/src/main/java/org/apache/arrow/performance/benchmark/micro/Float8Benchmarks.java]: Revised accordingly. Thanks for your good suggestion.

**Comment by emkornfield** [comment]
Thanks. +1, LGTM

**Comment by emkornfield** [comment]
(test failure appears unrelated)

**Pull request closed without merging**